salt/states/smartos.py

```diff
@@ -0,0 +1,805 @@
+          - alias: test
+          - quota: 5
+          - max_physical_memory: 512
+          - nics:
+            -
+              - mac: "82:1b:8e:49:e9:12"
+              - nic_tag: trunk
+              - mtu: 1500
+              - ips: [ "dhcp" ]
+              - vlan_id: 10
+            -
+              - mac: "82:1b:8e:49:e9:13"
+              - nic_tag: trunk
+              - mtu: 1500
+              - ips: [ "dhcp" ]
+              - vlan_id: 30
+
+    cleanup_images:
+      smartos.image_vacuum
+
+    .. note::
+
+        Keep in mind that when removing entries from vmconfig they will not get removed from the vm's current configuration,
+         for nics, disk, tags, ... they get removed as via add_*, set_*, update_*, and remove_*. Other properties must be
+         manual reset to there default value just like when using vmadm update.
+
+'''
+from __future__ import absolute_import
+
+# Import Python libs
+import logging
+import os
+
+# Import Salt libs
+import salt.utils
+import salt.utils.files
+from salt.utils.odict import OrderedDict
+
+log = logging.getLogger(__name__)
+
+# Define the state's virtual name
+__virtualname__ = 'smartos'
+
+
+def __virtual__():
+    '''
+    Provides smartos state provided for SmartOS
+    '''
+    return __virtualname__ if 'vmadm.create' in __salt__ else False
+
+
+def _load_config():
+    '''
+    Loads and parses /usbkey/config
+    '''
+    config = {}
+
+    if os.path.isfile('/usbkey/config'):
+        with open('/usbkey/config', 'r') as config_file:
+            for optval in config_file:
+                if optval[0] == '#':
+                    continue
+                if '=' not in optval:
+                    continue
+                optval = optval.split('=')
+                config[optval[0].lower()] = optval[1].strip()
+    log.debug('read /usbkey/config: {0}'.format(config))
+    return config
+
+
+def _write_config(config):
+    '''
+    writes /usbkey/config
+    '''
+    with open('/usbkey/config.salt', 'w') as config_file:
+        config_file.write("#\n# This file was generated by salt\n#\n")
+        for prop in OrderedDict(sorted(config.items())):
+            config_file.write("{0}={1}\n".format(prop, config[prop]))
+
+    if os.path.isfile('/usbkey/config.salt'):
+        try:
+            salt.utils.files.rename('/usbkey/config.salt', '/usbkey/config')
+        except IOError:
+            return False
+        log.debug('wrote /usbkey/config: {0}'.format(config))
+        return True
+    else:
+        return False
+
+
+def _parse_state_config(config, default_config):
+    '''
+    Parse vm_present state config
+    '''
+    for cfg_item in config if config else []:
+        for k in cfg_item:
+            if k in default_config:
+                if isinstance(cfg_item[k], (bool)):
+                    default_config[k] = cfg_item[k]
+                else:
+                    log.warning('config property {0} must be bool, using default: {1}'.format(k, default_config[k]))
```

Comment:
We might want to indicate which  function these warnings are coming from. Otherwise, this will be very hard for a user to track down.